### PR TITLE
Add support to redirect table reads from Hive to Iceberg

### PR DESCRIPTION
**Pull request by phd3**
Hive plugin changes based off of #7016 

When enabled, Hive Connector would redirect table reads to an Iceberg
catalog when it is an Iceberg table.

Co-authored by: Xingyuan Lin <linxingyuan1102@gmail.com>

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java [hunk 1]
```diff
@@ -146,6 +147,9 @@
 
     private boolean legacyHiveViewTranslation;
 
+    private boolean redirectIcebergTables;
+    private String targetCatalogForIcebergTables = "iceberg";
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:151]: Do we need two configs?
Maybe just
```
hive.iceberg-catalog-name
```

?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 2]
```diff
@@ -329,6 +338,12 @@
     private final HiveMaterializedViewMetadata hiveMaterializedViewMetadata;
     private final AccessControlMetadata accessControlMetadata;
 
+    // Copied from IcebergTableHandle
+    private static final Pattern ICEBERG_TABLE_NAME_PATTERN = Pattern.compile("" +
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:342]: _move_ it, so it stays in sync

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 3]
```diff
@@ -329,6 +338,12 @@
...
+            "(?<table>[^$@]+)" +
+            "(?:@(?<ver1>[0-9]+))?" +
+            "(?:\\$(?<type>[^@]+)(?:@(?<ver2>[0-9]+))?)?");
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:345]: I am not convinced we need to support any _implicit_ tables (snapshots, versions, or system tables like hive's $partitions).

I think the primary objective of this feature is "allow user to query whatever they see in SHOW TABLES".
and it's **not** to "provide a hiveberg catalog that merges hive's and iceberg's functionalities"

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 4]
```diff
@@ -3068,6 +3095,36 @@ public MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession s
...
+            return false;
+        }
+        return ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(table.get().getParameters().get(TABLE_TYPE_PROP));
+    }
+
+    private void rejectRedirectedTables(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (redirectTable(session, tableName).isPresent()) {
+            throw new TrinoException(NOT_SUPPORTED, "Hive Connector doesn't support modification operations (write, DDL, comment, statistics collection, set authorization) " +
+                    "on Iceberg tables when redirection is enabled");
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3124]: Do you mean that redirects on modification operations are a bad idea?
then why do we support them at all?

if they are a good idea in general, why would we want to opt out here?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java [hunk 5]
```diff
@@ -99,6 +99,8 @@
     private static final String DYNAMIC_FILTERING_PROBE_BLOCKING_TIMEOUT = "dynamic_filtering_probe_blocking_timeout";
     private static final String OPTIMIZE_SYMLINK_LISTING = "optimize_symlink_listing";
     private static final String LEGACY_HIVE_VIEW_TRANSLATION = "legacy_hive_view_translation";
+    private static final String REDIRECT_ICEBERG_TABLES = "redirect_iceberg_tables";
+    private static final String TARGET_CATALOG_FOR_ICEBERG_TABLES = "target_iceberg_catalog";
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java:103]: One property should be enough

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java [hunk 6]
```diff
@@ -410,6 +412,14 @@ public HiveSessionProperties(
...
                         "Use legacy Hive view translation mechanism",
                         hiveConfig.isLegacyHiveViewTranslation(),
+                        false),
+                booleanProperty(REDIRECT_ICEBERG_TABLES,
+                        "Enable redirecting iceberg tables to a different catalog",
+                        hiveConfig.isRedirectIcebergTables(),
+                        false),
+                stringProperty(TARGET_CATALOG_FOR_ICEBERG_TABLES,
+                        "The target catalog for redirecting iceberg tables",
+                        hiveConfig.getTargetCatalogForIcebergTables(),
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSessionProperties.java:422]: what are security and usability implication of allow users to set this at runtime?
it would be safer to have this as a config property only

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java [hunk 7]
```diff
@@ -3082,8 +3104,11 @@ public void testIllegalStorageFormatDuringTableScan()
         // to make sure it can still be retrieved instead of throwing exception.
         try (Transaction transaction = newTransaction()) {
             ConnectorMetadata metadata = transaction.getMetadata();
-            Map<SchemaTableName, List<ColumnMetadata>> allColumns = metadata.listTableColumns(newSession(), new SchemaTablePrefix(schemaTableName.getSchemaName()));
-            assertTrue(allColumns.containsKey(schemaTableName));
+            assertEquals(
+                    metadata.streamTableColumns(newSession(), new SchemaTablePrefix(schemaTableName.getSchemaName()))
+                            .filter(result -> result.getTable().equals(schemaTableName))
+                            .count(),
+                    1);
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:3111]: Use `assertThat` so that, should this fail on CI one day, the exception message produced would be uiseful

**Review by findepi — CHANGES_REQUESTED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java (hunk 1 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:151]: +1
Maybe the boolean config can be just a session property to allow easy disabling for a particular query

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java [hunk 8]
```diff
@@ -2900,16 +2913,25 @@ public void testHideDeltaLakeTables()
                         .doesNotContain(tableName);
 
                 // list all columns
-                assertThat(metadata.listTableColumns(session, new SchemaTablePrefix()).keySet())
-                        .doesNotContain(tableName);
+                assertEquals(
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:2916]:
```
assertThat(metadata.streamTableColumns(session, new SchemaTablePrefix()))
                        .extracting(TableColumnsMetadata::getTable)
                        .doesNotContain(tableName)
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java [hunk 9]
```diff
@@ -3082,8 +3104,11 @@ public void testIllegalStorageFormatDuringTableScan()
         // to make sure it can still be retrieved instead of throwing exception.
         try (Transaction transaction = newTransaction()) {
             ConnectorMetadata metadata = transaction.getMetadata();
-            Map<SchemaTableName, List<ColumnMetadata>> allColumns = metadata.listTableColumns(newSession(), new SchemaTablePrefix(schemaTableName.getSchemaName()));
-            assertTrue(allColumns.containsKey(schemaTableName));
+            assertEquals(
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:3107]:
```
assertThat(metadata.streamTableColumns(newSession(), new SchemaTablePrefix(schemaTableName.getSchemaName())))
                    .extracting(TableColumnsMetadata::getTable)
                    .containsOnlyOnce(schemaTableName)
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+        onTrino().executeQuery("DROP TABLE iceberg.default.test_alter_table");
+    }
+
+    @Test(groups = {HIVE_REDIRECTION_TO_ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testCommentTableFailure()
+    {
+        onTrino().executeQuery(format(CREATE_ICEBERG_TABLE_TEMPLATE, "iceberg.default.test_comment_table"));
+
+        assertNull(extractTableComment((String) onTrino().executeQuery("SHOW CREATE TABLE hive.default.test_comment_table").row(0).get(0)));
```
Comment (raunaqmorarka) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:130]: You can use `system.metadata.table_comments` to get the comment more easily

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:151]: why would you want disable/enable this?
what use-cases are we concerned about?

(bottom line is: we should have as much flexibility as required, and no more)

Comment (phd3) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:151]: IIRC there was an agreement to use two, e.g. https://github.com/trinodb/trino/pull/4704#discussion_r477567527. But can't find previous discussion on it. I like the idea of just having one config. 

@findepi One usecase for disabling would be to easily switch back to hive if 
(1) metadata is also available in Hive and 
(2) someone would like to temporarily fallback to hive instead of iceberg if iceberg scans fails or is suboptimal for a particular query shape. Just removing `table_type=iceberg` from HMS is sufficient but not necessary for all queries and engines. 

I understand that it's still very broken to do this, but can be useful during hive->iceberg transition.

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:151]: > ) someone would like to temporarily fallback to hive instead of iceberg if iceberg scans fails or is suboptimal for a particular query shape. Just removing table_type=iceberg from HMS is sufficient but not necessary for all queries and engines.

then the redirect wouldn't happen anyway

(and user would not expect correct results, so feels like a hack)

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:151]: Anyway, with single config it's equally easy to disable redirects -- just remove the value of the config.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 4 above)
Comment (phd3) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: @findepi

> Do you mean that redirects on modification operations are a bad idea?

https://github.com/trinodb/trino/pull/7606#issuecomment-843542294

does it look better with `Fail modifications for redirected tables in engine` commit? If so, I can put it as the first commit. Note that Hive connector check is still required because of procedures.

**Comment by phd3** [comment]
Unrelated #8477 and #8345

**Comment by phd3** [comment]
@findepi @losipiuk @raunaqmorarka @electrum this is ready for review.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: > does it look better with `Fail modifications for redirected tables in engine` commit?

I am not very fluent with this code, but if this implements @electrum 's thinking https://github.com/trinodb/trino/pull/7606#issuecomment-843542294, it should go in separate PR and to be debated separately. There hopefully is nothing special about redirects in Iceberg or Hive connector to warrant DDL-specific checks in the connector code (except, maybe, procedures -- which ones?)

Comment (phd3) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: makes sense to do it in a separate PR, was trying to gather feedback if exceptions thrown in hive redirection tests feel more natural with this change. and yes, that commit doesn't have anything specific to hive connector. 

hive connector checks become sort of "illegal state checks" in non-procedure calls, but wouldn't hurt to still keep proper error message there.

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: > but wouldn't hurt to still keep proper error message there.

except that it suggests -- to the future reader -- that we're coding some connector-specific behavior.

(and I do care about future readers quite a lot:)

Comment (phd3) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Good point. How about changing all connector checks to `checkState`, and adding special handling for procedures that throw TrinoException?

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: or TrinoException here, but with a comment. up to you

**Comment by findepi** [comment]
@phd3 can you please rebase?

File: core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java [hunk 11]
```diff
@@ -72,7 +72,7 @@ public String getName()
     {
         Session session = stateMachine.getSession();
         QualifiedObjectName tableName = createQualifiedObjectName(session, statement, statement.getName());
-        Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
+        Optional<TableHandle> tableHandle = metadata.getOriginalTableHandle(session, tableName, Optional.of(getName()));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:75]: > Fail modifications for redirected tables in engine

I am not convinced we should to that

Comment (phd3) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:75]: here "modifications" is bad wording, meant DDLs. we already added support for DMLs in https://github.com/trinodb/trino/pull/8683

The last discussion that we had on this was https://github.com/trinodb/trino/pull/7606#issuecomment-843542294

> DDL operations are problematic since users need to be aware of which connector they are using, since they have different data types, partitioning, bucketing, etc. Our thought was that these are relatively rare operations by more advanced users and that trying to have hidden redirections would end up causing more confusion.

Comment (phd3) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:75]: do you also feel otherwise for DDLs?

Comment (phd3) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:75]: do you also feel otherwise for DDLs?

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java:75]: Users sooner or later will ask why `ALTER ... ADD COLUMN (something varchar)` does not work, and I won't be able to explain to them why not. Yes, from engineer perspective, we may have harder time supporting things like column properties (maybe, or maybe not), but i would assume there are not always used, so users who do not intend to use column properties, will not accept this as a rational explanation for the limitation.
TL;DR yes, DDLs like ALTER .. ADD/DROP COLUMN should be routed as well.

cc @losipiuk @alexjo2144

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 12]
```diff
@@ -2996,6 +3021,37 @@ public boolean delegateMaterializedViewRefreshToConnector(ConnectorSession sessi
         return hiveMaterializedViewMetadata.refreshMaterializedView(session, name);
     }
 
+    @Override
+    public Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName)
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3025]: Can you please update this to https://github.com/trinodb/trino/pull/9870?

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3025]: > Can you please update this to #9870?

I've taken liberty to update the code myself. 
I also copied & adapted an awesome test template done by @ssheikin too.
posted https://github.com/trinodb/trino/pull/10173 with the changes

Comment (ssheikin) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3025]: This credits belong to @MiguelWeezardo

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3025]: > These credits belong to @MiguelWeezardo

thanks @ssheikin for the comment. added @MiguelWeezardo as co-author in the other PR

**Comment by raunaqmorarka** [comment]
Seems to be superseded by https://github.com/trinodb/trino/pull/10173
